app/main/views/questionnaire.py

```diff
@@ -47,6 +48,55 @@ def add_cache_control(response):
     return response
 
 
+@questionnaire_blueprint.route('<group_id>/<int:group_instance>/<block_id>', methods=["GET"])
+@login_required
+def get_block(eq_id, form_type, collection_id, group_id, group_instance, block_id):
+    # Filter answers down to those we may need to render
+    answer_store = get_answer_store(current_user)
+    answers = answer_store.map(group_id=group_id, group_instance=group_instance, block_id=block_id)
+
+    q_manager = get_questionnaire_manager(g.schema, g.schema_json)
+    q_manager.build_state(block_id, answers)
+
+    block = g.schema.get_item_by_id(block_id)
+    template = block.type if block and block.type else 'questionnaire'
+
+    return _render_template(block_id, q_manager.state, template=template)
+
+
+@questionnaire_blueprint.route('<group_id>/<int:group_instance>/<block_id>', methods=["POST"])
+@login_required
+def post_block(eq_id, form_type, collection_id, group_id, group_instance, block_id):
+    navigator = Navigator(g.schema_json, get_answer_store(current_user))
+    q_manager = get_questionnaire_manager(g.schema, g.schema_json)
+
+    this_block = {
+        'group_id': group_id,
+        'group_instance': group_instance,
+        'block_id': block_id,
+    }
+
+    valid_location = this_block in navigator.get_location_path()
+    valid_data = q_manager.validate(block_id, request.form)
+
+    if not valid_location or not valid_data:
+        return _render_template(block_id, q_manager.state, template='questionnaire')
+    else:
+        q_manager.update_questionnaire_store(this_block)
+
+    next_location = navigator.get_next_location(current_group_id=group_id, current_iteration=group_instance, current_block_id=block_id)
+
+    metadata = get_metadata(current_user)
+
+    next_block_id = None if next_location is None else next_location['block_id']
```

Comment:
I think None is actually a valid response for previous/next links where we're looking at introduction or a summary page say. The template isn't using them, but the render method require's they're specified.